### PR TITLE
Bag rewriter (C++)

### PR DESCRIPTION
**Pull request by emersonknapp**
Implements main functionality of #831
Depends on #924 
Depends on #923 

Adds a `bag_rewrite` API that takes Reader(s) for existing bags and writes all their messages to Writer(s) with specified `StorageOptions` and `RecordOptions`.

Does not expose the feature to Python/CLI - that is implemented in followup PR #921

### CODE REVIEW

**Review by jhdcs — COMMENTED** [review]
Only seeing one really small comment issue (apart from the build issue). Not sure if it's something we need to bother with, but thought I'd bring it up anyways.

File: rosbag2_transport/src/rosbag2_transport/bag_rewrite.cpp [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+      earliest_msg = msg;
+      earliest_msg_index = i;
+    }
+  }
+
+  // clear returned message from queue before returning it, so it can be refilled next time
+  if (earliest_msg != nullptr) {
+    next_messages[earliest_msg_index].reset();
+  }
+  return earliest_msg;
```
[the remark this answers is not in the document]
Comment (lihui815) [reply rosbag2_transport/src/rosbag2_transport/bag_rewrite.cpp:65]: That means we're done reading all input messages right? I think returning `nullptr` is pretty reasonable behavior in that case. This is also only used as a helper for `perform_rewrite` which as far as I can tell is handling it properly.

File: rosbag2_transport/src/rosbag2_transport/bag_rewrite.cpp [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+    // Done filtering - set up writer
+    for (const auto & [topic_name, topic_type] : filtered_topics_and_types) {
+      rosbag2_storage::TopicMetadata topic_metadata;
+      topic_metadata.name = topic_name;
+      topic_metadata.type = topic_type;
+
+      // Take source serialization format for the topic if output format is unspecified
+      if (record_options.rmw_serialization_format.empty()) {
+        topic_metadata.serialization_format = input_topics_serialization_format[topic_name];
```
Comment (lihui815) [inline rosbag2_transport/src/rosbag2_transport/bag_rewrite.cpp:116]: In the rare case that the same topic has different serialization formats in the inputs, this is the "last" serialization format that occurs in the previous loop right? Maybe note somewhere that this is the expected selection behaviour?

**Review by lihui815 — APPROVED** [review]
looks reasonable to me. just one nit on documenting serialization format selection.

File: rosbag2_transport/src/rosbag2_transport/bag_rewrite.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (jhdcs) [reply rosbag2_transport/src/rosbag2_transport/bag_rewrite.cpp:65]: Fair enough. I must have been mis-reading something there for a moment!

File: rosbag2_transport/src/rosbag2_transport/bag_rewrite.cpp (hunk 2 above)
Comment (emersonknapp) [reply rosbag2_transport/src/rosbag2_transport/bag_rewrite.cpp:117]: Added a note comment in the docstring in the header.

File: rosbag2_transport/src/rosbag2_transport/bag_rewrite.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (emersonknapp) [reply rosbag2_transport/src/rosbag2_transport/bag_rewrite.cpp:66]: Added a note to the docstring for this function that it returns nullptr when all inputs are done reading. `perform_rewrite` does indeed notice that output and return.

**Comment by emersonknapp** [comment]
Gist: https://gist.githubusercontent.com/emersonknapp/118ff3d83fe59438cad609306b4c102d/raw/8368bbbc8a74f487474f3474161702962fbd4cb6/ros2.repos
BUILD args: --packages-up-to rosbag2_transport rosbag2_tests rosbag2
TEST args: --packages-select rosbag2_transport rosbag2_tests rosbag2
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/9433
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15707/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10389/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/13374/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15949/)

**Comment by emersonknapp** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15712/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10394/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/13377/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15954/)

**Comment by emersonknapp** [comment]
Noting here explicitly. The OSX CI build queue is backed up by ~12 hours right now because there is only a single executor online and some very long builds are stacked up. Will merge now, and if something is warning on OSX then I am happy to fix it then.

**Comment by clalancette** [comment]
> Noting here explicitly. The OSX CI build queue is backed up by ~12 hours right now because there is only a single executor online and some very long builds are stacked up.

Whoops.  We were doing some network maintenance and we forgot to reenable them when that was finished.  All 4 are back online now, FYI.

**Comment by emersonknapp** [comment]
Looks like the one from this morning failed on some networking issue, and the one from 3 hours ago got remove from the queue somehow (Jenkins doesn't have actual entities for pending builds, which drives me bananas - if you delete your build from the queue then everybody else's badges from behind you in the queue are wrong because the ID gets reassigned to the next queue item </rant>)

Rebuilding for just in case, though #921 will catch any problems

OSX: [Build Status](https://ci.ros2.org/job/ci_osx/13379/)

File: rosbag2_transport/src/rosbag2_transport/bag_rewrite.cpp [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    throw std::runtime_error("Must provide at least one input and one output bag to rewrite.");
+  }
+
+  auto topic_outputs = setup_topic_filtering(input_bags, output_bags);
+
+  std::vector<std::shared_ptr<rosbag2_storage::SerializedBagMessage>> next_messages;
+  next_messages.resize(input_bags.size(), nullptr);
+
+  std::shared_ptr<rosbag2_storage::SerializedBagMessage> next_msg;
+  while (next_msg = get_next(input_bags, next_messages)) {
```
Comment (Blast545) [inline rosbag2_transport/src/rosbag2_transport/bag_rewrite.cpp:152]: This introduced a warning message into the buildfarm. See [nightly_osx_debug#2246](https://ci.ros2.org/view/nightly/job/nightly_osx_debug/2246/).

It's probably unintended usage, would you take a look? @emersonknapp

**Comment by Blast545** [comment]
:man_farmer: This test regression is probably related to this PR as well: https://ci.ros2.org/view/nightly/job/nightly_win_rel/2137/

**Comment by emersonknapp** [comment]
@Blast545 no, that is an existing flaky test - there is no way for this PR to have affected recording functionality

File: rosbag2_transport/src/rosbag2_transport/bag_rewrite.cpp (hunk 3 above)
Comment (emersonknapp) [reply rosbag2_transport/src/rosbag2_transport/bag_rewrite.cpp:152]: It is intended usage - as noted above I wasn't 100% sure that OSX would be happy. Will add the parentheses suggested by Clang ASAP.

Comment (emersonknapp) [reply rosbag2_transport/src/rosbag2_transport/bag_rewrite.cpp:152]: #930

**Pull request merged**